lib/ts/controllers/s-modal.ts

```diff
@@ -0,0 +1,222 @@
+namespace Stacks {
+
+    export class ModalController extends Stacks.StacksController {
+        static targets = ["modal"];
+
+        private modalTarget!: HTMLElement;
+
+        private _boundClickFn!: any;
+        private _boundKeypressFn!: any;
+
+        private returnElement!: HTMLElement;
+
+        private _boundTabTrap!: any;
+
+        connect () {
+            this._validate();
+        }
+
+        /**
+         * Disconnects all added event listeners on controller disconnect
+         */
+        disconnect() {
+            this._unbindDocumentEvents();
+        };
+
+        /**
+         * Toggles the visibility of the modal
+         */
+        toggle () {
+            this._toggle();
+        }
+
+        /**
+         * Shows the modal
+         */
+        show () {
+            this._toggle(true);
+        }
+
+        /**
+         * Hides the modal
+         */
+        hide () {
+            this._toggle(false);
+        }
+
+        /**
+         * Validates the modal settings and attempts to set necessary internal variables
+         */
+        _validate() {
+            // check for returnElement support
+            var returnElementSelector = this.data.get("return-element");
+            if (returnElementSelector) {
+                this.returnElement = <HTMLElement>document.querySelector(returnElementSelector);
+
+                if (!this.returnElement) {
+                    throw "Unable to find element by return-element selector: " + returnElementSelector;
+                }
+            }
+        }
+
+        /**
+         * Toggles the visibility of the modal element
+         * @param show Optional parameter that force shows/hides the element or toggles it if left undefined
+         */
+        _toggle (show?: boolean | undefined) {
+            var toShow = show;
+
+            // if we're letting the class toggle, we need to figure out if the popover is visible manually
+            if (typeof toShow === "undefined") {
+                toShow = this.modalTarget.getAttribute("aria-hidden") === "true";
+            }
+
+            // show/hide events trigger before toggling the class
+            this.triggerEvent(toShow ? "show" : "hide");
+
+            this.modalTarget.setAttribute("aria-hidden", toShow ? "false" : "true");
+
+            if (toShow) {
+                this._bindDocumentEvents();
+            }
+            else {
+                this._unbindDocumentEvents();
+                this._focusReturnElement();
```

Comment:
Situations on the ground may vary.  The comment flagging dialog could could have a return element inside the comment row but then the comment row gets deleted.

We should do something along the lines of the following, so clients can adjust the return element on `s-modal:hide`:

```
var event = this.triggerEvent("hide", { returnElement: this.returnElement });
this.modalTarget.setAttribute("aria-hidden", "true");
this._unbindDocumentEvents();
this._focusReturnElement(event.data.returnElement);

Reply:
See above comment. I'm going to take your earlier advice to add support for setting the return element in `show/hideModal` instead since that seems like a more friendly implementation to me